required/firstaid/latex2e-first-aid-for-external-files.dtx

```diff
@@ -655,6 +655,69 @@
 }
 %    \end{macrocode}
 %
+% \subsection{The \pkg{cleveref} package first aid}
+%
+%    The \pkg{cleveref} package expects only two data containers
+%    for its internal \cs{newlabel} command. This fails if
+%    \pkg{xr-hyper} is used which expands every \cs{newlabel} to 
+%    five data container and puts the file name into the last one.
+%    \begin{macrocode}
+\AddToHook{package/cleveref/after}[firstaid]{%
+  \FirstAidNeededT{cleveref}{sty}{2018/03/27 v0.21.4 Intelligent cross-referencing}
+     {%
+%    \end{macrocode}
+% This are the two commands which retrieve the data from the label info.
+% We change it to expand five arguments.
+%    \begin{macrocode}
+       \def\cref@getref#1#2{%
+         \expandafter\let\expandafter#2\csname r@#1@cref\endcsname%
+         \expandafter\expandafter\expandafter\def%
+           \expandafter\expandafter\expandafter#2%
+           \expandafter\expandafter\expandafter{%
+             \expandafter\@firstoffive#2}}% <-------- five
+       \def\cpageref@getref#1#2{%
+         \expandafter\let\expandafter#2\csname r@#1@cref\endcsname%
+         \expandafter\expandafter\expandafter\def%
+           \expandafter\expandafter\expandafter#2%
+           \expandafter\expandafter\expandafter{%
+             \expandafter\@secondoffive#2}}% <----------- five  
+%    \end{macrocode}
+%  We also need to redefine the internal label commands of cleveref.
+%  This must be done after cleveref, so we use an unique label.
```

Comment:
not sure I understand this: without a rule I don't think this is guaranteed to follow what cleveref puts into begindocument (it will initially, but ...).

I would guess you either have to do \RemoveFromHook{begindocument}[cleveref] or you use the cleveref label to add it in which case it should be appended.

Reply:
I would have expected that the order stays, but you are quite right, it makes much sense to use the existing cleveref label and to append to this chunk. (You can't remove first, there is other code in the hook which must stay)

Reply:
you are right, it uses \AtBeginDocument many times ... I didn't check for that. So yes, appending to it is the right thing to do (there the order is well-defined)